echo-notifications/src/test/groovy/com/netflix/spinnaker/echo/notification/GoogleCloudBuildNotificationSpec.groovy

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.notification
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.echo.model.Event
+import com.netflix.spinnaker.echo.model.Metadata
+import com.netflix.spinnaker.echo.model.pubsub.MessageDescription
+import com.netflix.spinnaker.echo.services.IgorService
+import com.netflix.spinnaker.kork.core.RetrySupport
+import spock.lang.Specification
+import spock.lang.Subject
+
+class GoogleCloudBuildNotificationSpec extends Specification {
+  IgorService igorService = Mock(IgorService)
+  RetrySupport retrySupport = new RetrySupport()
+  ObjectMapper objectMapper = new ObjectMapper()
+
+  final String ACCOUNT_NAME = "my-account"
+  final String BUILD_ID = "1a9ea355-eb3d-4148-b81b-875d07ea118b"
+  final String BUILD_STATUS = "QUEUED"
+  final String PAYLOAD = objectMapper.writeValueAsString([
+    "key": "value"
+  ])
+
+  @Subject
+  GoogleCloudBuildNotificationAgent notificationAgent = new GoogleCloudBuildNotificationAgent(igorService, retrySupport)
+
+  def "ignores non-googleCloudBuild events"() {
+    given:
+    Event event = createEvent("pubsub")
+
+    when:
+    notificationAgent.processEvent(event)
+
+    then:
+    0 * igorService._
+  }
+
+  def "sends googleCloudBuild events to igor"() {
+    given:
+    Event event = createEvent("googleCloudBuild")
+
+    when:
+    notificationAgent.processEvent(event)
+
+    then:
+    1 * igorService.updateBuildStatus(ACCOUNT_NAME, BUILD_ID, BUILD_STATUS, { it -> it.in().text == PAYLOAD })
+    0 * igorService._
+  }
+
+  def "retries on failure to communicate with igor"() {
+    given:
+    Event event = createEvent("googleCloudBuild")
+
+    when:
+    notificationAgent.processEvent(event)
+
+    then:
+    2 * igorService.updateBuildStatus(ACCOUNT_NAME, BUILD_ID, BUILD_STATUS, { it -> it.in().text == PAYLOAD }) >>
```

Comment:
What is the intended behavior of duplicate build statuses (statii)?

Reply:
You mean if we receive a notification twice?  We'll forward along both to igor.   Igor will refuse to update its cache with an older build status (to prevent issues due to out-of-order messages) but will allow a new message with the same status.